src/verifier.py

```diff
@@ -22,9 +22,35 @@ def verify_ballot_consistency(self) -> bool:
         the "Proving Equality of Arrays of Vote Values" procedure described
         in Section II-F.
         """
-        # TODO
-        # Parse SBB
-        # Use (t, -t) and prior commitments posted to verify
+        sbb_contents = self._sbb.get_sbb_contents()
+        
+        # First, validate the commitment consistency with the initial vote lists and final vote lists.
+        for list_idx, proof in sbb_contents.consistency_proof.items():
+            for vote_idx in range(len(proof)):
+                proved_sv = proof[vote_idx]
+                for row_idx, sv in enumerate(proved_sv):
+                    # Ensure that we are consistent with the initial and the final commitments
+                    if sv.get('u', None) is not None:
+                        val_init = sv['u_init']
+                        val_fin = sv['u_fin']
+                        original_commitment = sbb_contents.svr_commitments[row_idx][vote_idx]['com_u']
+                        final_commitment = sbb_contents.vote_lists[list_idx][vote_idx][row_idx].com_u
+                    else:
+                        val_init = sv['v_init']
+                        val_fin = sv['v_fin']
+                        original_commitment = sbb_contents.svr_commitments[row_idx][vote_idx]['com_v']
+                        final_commitment = sbb_contents.vote_lists[list_idx][vote_idx][row_idx].com_v
+                    key_init = sv['k_init']
+                    key_fin = sv['k_fin']
+                    
+                    com_init = util.get_COM(util.bigint_to_bytes(key_init), util.bigint_to_bytes(val_init))
+                    com_fin = util.get_COM(util.bigint_to_bytes(key_fin), util.bigint_to_bytes(val_fin))
+                    if com_init != original_commitment:
+                        raise Exception("Failed to open the initial vote commitment")
+                    if com_fin != final_commitment:
+                        raise Exception("Failed to open the final vote commitment")
+            
+        # TODO: Use (t, -t) and prior commitments posted to verify consistency via lagrange
```

Comment:
I saw the use of Lagrange Interpolation as well, but it doesn't make much sense to me. Do you understand it or are you just referencing that in the Rivest impl?

I read this section over again in the paper and it seems all you need to do is the `t` calculation from Section II-E once for each row, and if they all are equal then that vote is valid? But then this doesn't actually work when I tried adding it to your branch:

```
        for list_idx, proof in sbb_contents.consistency_proof.items():
            for vote_idx in range(len(proof)):
                proved_sv = proof[vote_idx]
                for row_idx, sv in enumerate(proved_sv):
                    if 'u' in sv:
                        tu = sbb_contents.t_values[list_idx][row_idx][vote_idx]['tu']
                        if tu != (sv['u_fin'] - sv['u_init']) % self._M:
                            raise Exception("Failed vote consistency proof")
                    else:
                        tv = sbb_contents.t_values[list_idx][row_idx][vote_idx]['tv']
                        if tv != (sv['v_fin'] - sv['v_init']) % self._M:
                            raise Exception("Failed vote consistency proof")
```

Any idea why not? Is my understanding wrong?

Reply:
nope, I don't have a good understanding of it, and it's also very possible that my t values are just wrong. I'll be taking a look at it today

Reply:
I added verification of t-values and it works as expected. The snippet above needs to use `sv['u']` instead of `sv['u_fin']`

I think the misunderstanding is that `u` and `u_fin` are the same, except `u_fin` is in the order of the final output commitments. Instead of unmixing using the permutations in the verifier, I just publish the values in the unmixed order, which makes it easier to read (at least for me).

Reply:
I added the lagrange interpolation code directly from rivest, but wasn't able to get it working, and have no clue where to even start with debugging